### ISSUE TITLE
Great work! Couple of small notes

### ISSUE THREAD

**Issue by jordansavant** [jakebesworth/Simple-SDL2-Audio#10]
First off I want to say this library was a lifesaver and I am very impressed by it! Good work. I had a couple of notes after its usage I wanted to bring up.

### 1. I followed all implementation notes and details but could not get sound or music to play on my Windows 10 laptop (2021 Razor Blade).

After rolling up my sleeves and debugging the `audio.c` I narrowed it down to this code:
```c
if((gDevice->device = SDL_OpenAudioDevice(NULL, 0, &(gDevice->want), &have, SDL_AUDIO_ALLOW_ANY_CHANGE)) == 0)
```
and
```c
SDL_CloseAudioDevice(gDevice->device);
```

By commenting out those two lines and replacing them with the default SDL versions the audio began working as expected.
```c
if (SDL_OpenAudio(&(gDevice->want), NULL) < 0 )
```
and
```c
SDL_CloseAudio();
```
I reveiwed the SDL2 documentation and was able to confirm that it was indeed using the "default" device (laptop speakers) but as their documentation indicated the device assigned was never `0` for backwards compatibility purposes. I am not sure that really meant anything for helping diagnose the issue and perhaps it is a bug in SDL itself for my version of the build (Windows 10 32bit libs)

*edit* just to be clear I did indeed try:
`#define SDL_AUDIO_ALLOW_CHANGES SDL_AUDIO_ALLOW_FREQUENCY_CHANGE | SDL_AUDIO_ALLOW_CHANNELS_CHANGE`

### 2. I extended with a `type` so I could control loop and volumes separately for music or sounds

I wanted to play music one time and then stop and also be able to potentially loop sound effects. I also wanted separate volumes for both sounds and music. To solve this I extended the implementation with an `enum AudioType` and put the type on the `Audio` struct. Then when I called playAudio I specified the type and treated it as appropriate as music or sounds.

This involved extending `type` checks in the `stopMusic`, `addMusic` and `audioCallback` to appropriately fade out music types and decrement sound effects.

I figured there was a design choice for using `loop` to specify the type and didn't want to waste your time with a pull request for something like that. If you do want to just let me know.

Again thanks for this.

**Comment by jakebesworth** [comment]
Hey @jordansavant ! Thanks so much for reaching out with your findings.

First question: What version of SDL2 are you using?

`SDL_OpenAudio: This function is a legacy means of opening the audio device.` This is the older way of opening audio devices which is why it wasn't used, but also why it might work for you.

They both look like this:

```
SDL_AudioDeviceID SDL_OpenAudioDevice(
                          const char *device,
                          int iscapture,
                          const SDL_AudioSpec *desired,
                          SDL_AudioSpec *obtained,
                          int allowed_changes);

// device | a UTF-8 string reported by SDL_GetAudioDeviceName() or a driver-specific name as appropriate. NULL requests the most reasonable default device.
// iscapture | non-zero to specify a device should be opened for recording, not playback
// desired | an SDL_AudioSpec structure representing the desired output format; see SDL_OpenAudio() for more information
// obtained | an SDL_AudioSpec structure filled in with the actual output format; see SDL_OpenAudio() for more information
// allowed_changes | 0, or one or more flags OR'd together
```

```
int SDL_OpenAudio(SDL_AudioSpec * desired,
                  SDL_AudioSpec * obtained);

// desired | an SDL_AudioSpec structure representing the desired output format. Please refer to the SDL_OpenAudioDevice documentation for details on how to prepare this structure.
// obtained | an SDL_AudioSpec structure filled in with the actual parameters, or NULL.
```

Since they are so similar, you may want to change the parameters of `SDL_GetAudioDeviceName` such as setting `SDL_AudioSpec *obtained` to `NULL` and seeing if you can narrow down the specific difference.

> I figured there was a design choice for using loop to specify the type and didn't want to waste your time with a pull request for something like that. If you do want to just let me know.

I added loop just to allow music, and sounds. It was a basic use case. In the future I hoped to add ability to stop, and start specific audio, ability to change volume, and other features. Since this project is used so widely a pull request for this feature would need to be somewhat backwards compatible (so a simple addition, not adding too much complexity to the project). If you don't mind making some changes based on my opinions, feel free to open a pull request and we could work on such an addition!

**Comment by jordansavant** [comment]
Sorry it took me a bit to get back to this during the holidays. Let me research a little closer on your details and I will provide more information on the SDL2 version and test out your suggestions in the next few days.

**Comment by jordansavant** [comment]
So I took a further look and I tested it on `2.0.16` and `2.0.18`. I was able to replicate the failure with the older `SDL_OpenAudio` method by passing in the `SDL_AudioSpec *obtained` property that was pulled from a wav file. Their documentation stated:

> If obtained is NULL, the audio data passed to the callback function will be guaranteed to be in the requested format, and will be automatically converted to the actual hardware audio format if necessary. If obtained is NULL, desired will have fields modified.

Since I was not passing NULL to the old method I assume it did not change formats to meet the hardware and probably something is similarly happening on the newer method `SDL_OpenAudioDevice`. As it states in the remarks of the newer method:

> If obtained is NULL, we use desired (and allow no changes), which means desired will be modified to have the correct values for silence, etc, and SDL will convert any differences between your app's specific request and the hardware behind the scenes

It says that when passing in obtained as NULL in `SDL_OpenAudioDevice` "we use desired (and allow no changes)" and I can only surmise that the "allow no changes" means its not converting the desired to the hardware probably. Regardless it won't play with passing in an `obtained` either.

Long story short, I was able to replicate all of this without any of the Simple-SDL2-Audio involved so it is purely a problem between SDL 2 and my machine. I will go ahead and close this issue.

Thank you!

**Comment by jakebesworth** [comment]
@jordansavant Thanks so much for the investigation! If you have the time, opening such an issue with the SDL community is relatively easy, and they would be glad to get some feedback on more niche cases.
